csrc/val_graph.cpp

```diff
@@ -0,0 +1,463 @@
+  std::vector<ValGroup> input_groups;
+  for (auto id_input : expr->front()->inputs()) {
+    input_groups.push_back(toGroup(id_input));
+  }
+  return input_groups;
+}
+
+ExprGroups ValGraph::allUsesOf(const ValGroups& of) const {
+  DequeOfExprGroup to_visit;
+  for (const ValGroup& of_val_group : of) {
+    if (const ExprGroups* group_uses = getUses(of_val_group);
+        group_uses != nullptr) {
+      to_visit.insert(to_visit.end(), group_uses->begin(), group_uses->end());
+    }
+  }
+
+  UnorderedSetOfExprGroup visited;
+  while (!to_visit.empty()) {
+    ExprGroup current_expr = to_visit.front();
+    to_visit.pop_front();
+    visited.emplace(current_expr);
+    for (const ValGroup& output_id : outputGroups(current_expr)) {
+      if (const ExprGroups* group_uses = getUses(output_id);
+          group_uses != nullptr) {
+        for (const ExprGroup& group_use : *group_uses) {
+          if (visited.count(group_use)) {
+            continue;
+          }
+          to_visit.push_back(group_use);
+        }
+      }
+    }
+  }
+
+  return visited;
+}
+
+ExprGroups ValGraph::allDefinitionsOf(const ValGroups& of) const {
+  DequeOfExprGroup to_visit;
+  for (const ValGroup& of_val_group : of) {
+    if (const ExprGroups* group_defs = getDefinitions(of_val_group);
+        group_defs != nullptr) {
+      to_visit.insert(to_visit.end(), group_defs->begin(), group_defs->end());
+    }
+  }
+
+  UnorderedSetOfExprGroup visited;
+  while (!to_visit.empty()) {
+    ExprGroup current_expr = to_visit.front();
+    to_visit.pop_front();
+    visited.emplace(current_expr);
+    for (const ValGroup& input_id : inputGroups(current_expr)) {
+      if (const ExprGroups* group_defs = getDefinitions(input_id);
+          group_defs != nullptr) {
+        for (const ExprGroup& group_def : *group_defs) {
+          if (visited.count(group_def)) {
+            continue;
+          }
+          to_visit.push_back(group_def);
+        }
+      }
+    }
+  }
+
+  return visited;
+}
+
+bool ValGraph::hasDefinitions(const ValGroup& val_group) const {
+  NVF_ERROR(val_group);
+  return unique_definitions_.find(val_group) != unique_definitions_.end();
+}
+
+bool ValGraph::hasUses(const ValGroup& val_group) const {
+  NVF_ERROR(val_group);
+  return unique_uses_.find(val_group) != unique_uses_.end();
+}
+
+std::string ValGraph::toString() const {
+  std::stringstream ss;
+  ss << "IdGraph { \n";
+  ss << "Disjoint Ids:\n"
+     << idGroupsString(*this, 1) << "\n\nDisjoint Expression groups:\n"
+     << exprGroupsString(*this, 1) << std::endl;
+  ss << " } IdGraph\n" << std::endl;
+  return ss.str();
+}
+
+void ValGraph::initializeVal(
+    Val* val,
+    const VectorOfUniqueEntries<Expr*>& definitions,
+    const VectorOfUniqueEntries<Expr*>& uses) {
+  const ValGroup& val_disjoint_set =
+      disjointValSets().initializeSet(val).first->second;
+
+  ExprGroups def_groups;
+  for (auto def : definitions) {
```

Comment:
If one val has multiple definitions, should they considered mapped? If for this PR, the definition is always unique, then should we add a `NVF_ERROR(definitions.size() == 1)`?

Reply:
Expr mapping needs both inputs and outputs being mapped, so obviously if two expressions generate the same val, the outputs are mapped, but that doesn't automatically mean the inputs are mapped too. I have no particular idea when this could happen, though.

Agree with the assertion. Added.